cpp/src/neighbors/detail/cagra/cagra_build.cuh

```diff
@@ -42,13 +42,1440 @@
+                 primary_file_offset,
+                 augmented_file_offset);
+
+  if (primary_size > 0) {
+    RAFT_LOG_DEBUG(
+      "ACE: Reading %lu primary vectors from offset %lu", primary_size, primary_file_offset);
+
+    const size_t primary_bytes = primary_size * vector_size;
+    ace_read_large_file<T, IdxT>(
+      reordered_fd, sub_dataset.data_handle(), primary_bytes, primary_file_offset);
+  }
+
+  if (augmented_size > 0) {
+    RAFT_LOG_DEBUG(
+      "ACE: Reading %lu augmented vectors from offset %lu", augmented_size, augmented_file_offset);
+
+    const size_t augmented_bytes = augmented_size * vector_size;
+    T* augmented_dest            = sub_dataset.data_handle() + (primary_size * dataset_dim);
+    ace_read_large_file<T, IdxT>(
+      augmented_fd, augmented_dest, augmented_bytes, augmented_file_offset);
+  }
+}
+
+// Build CAGRA index using ACE (Augmented Core Extraction) partitioning
+// ACE enables building indices for datasets too large to fit in GPU memory by:
+// 1. Partitioning the dataset using balanced k-means in primary (non-overlapping) and augmented
+// (second-closest) partitions
+// 2. Building sub-indices for each partition independently
+// 3. Concatenating sub-graphs (of primary partitions) into a final unified index
+// Supports both in-memory and disk-based modes depending on available host memory.
+// In disk mode, the graph is stored in ace_build_dir and dataset is reordered on disk.
+// The returned index is not usable for search. Use the created files for search instead.
+template <typename T, typename IdxT, typename Accessor>
+index<T, IdxT> build_ace(
+  raft::resources const& res,
+  const index_params& params,
+  raft::mdspan<const T, raft::matrix_extent<int64_t>, raft::row_major, Accessor> dataset,
+  size_t num_partitions = 0)
+{
+  common::nvtx::range<common::nvtx::domain::cuvs> function_scope(
+    "cagra::build_ace<%s>(%zu, %zu, %zu)",
+    Accessor::is_managed_type::value ? "managed"
+    : Accessor::is_host_type::value  ? "host"
+                                     : "device",
+    params.intermediate_graph_degree,
+    params.graph_degree,
+    num_partitions);
+
+  uint64_t dataset_size = dataset.extent(0);
+  uint64_t dataset_dim  = dataset.extent(1);
+
+  RAFT_EXPECTS(dataset_size > 0, "ACE: Dataset must not be empty");
+  RAFT_EXPECTS(dataset_dim > 0, "ACE: Dataset dimension must be greater than 0");
+  RAFT_EXPECTS(params.intermediate_graph_degree > 0,
+               "ACE: Intermediate graph degree must be greater than 0");
+  RAFT_EXPECTS(params.graph_degree > 0, "ACE: Graph degree must be greater than 0");
+
+  uint64_t n_partitions = (num_partitions > 0) ? num_partitions : params.ace_npartitions;
+  RAFT_EXPECTS(n_partitions > 0, "ACE: num_partitions must be greater than 0");
+  RAFT_EXPECTS(n_partitions <= dataset_size, "ACE: num_partitions cannot exceed dataset size");
+
+  uint64_t min_vectors_per_partition  = dataset_size / n_partitions;
```

Comment:
This is an avg number of vectors per partition, right? Would it be better to remove the `min_` prefix here?

Reply:
Good point. I have renamed it to `avg_vectors_per_partition`.